equinix/resource_metal_device_acc_test.go

```diff
@@ -1121,9 +1144,75 @@ func TestAccMetalDevice_readErrorHandling(t *testing.T) {
 			if tt.args.newResource {
 				d.MarkNewResource()
 			}
-			if err := resourceMetalDeviceRead(d, tt.args.meta); (err != nil) != tt.wantErr {
+			if err := resourceMetalDeviceRead(context.Background(), d, tt.args.meta); (err != nil) != tt.wantErr {
 				t.Errorf("resourceMetalDeviceRead() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})
 	}
 }
+
+func TestAccMetalDevice_timeouts(t *testing.T) {
+	d1 := &packngo.Device{
+		State: "active",
+	}
+	projSuffix := acctest.RandString(10)
+
+	mockMetalDeviceService := &mockDeviceService{
+		CreateFn: func(device *packngo.DeviceCreateRequest) (*packngo.Device, *packngo.Response, error) {
+			return d1, &packngo.Response{Response: &http.Response{Status: matchErrDeviceReadyTimeout.String(), StatusCode: 500}},
```

Comment:
I don't think we should mock packngo in order to return terraform provider error messages.  A mock that waits a certain amount of time before responding would be more accurate for the test case, but an integration test would be preferable; with a short timeout we should be able to guarantee a timeout error in that test.

Reply:
Shall we rename this to `TestMetalDevice_timeouts` (no `Acc`) so the offline test can run in our unit tests?

Reply:
I was overlooking that this adds more dependency on packngo (vs metal-go).

Reply:
If we want to stick with mocking _something_, this article lays out a few options that would allow us to mock API behavior rather than SDK behavior:

https://medium.com/zus-health/mocking-outbound-http-requests-in-go-youre-probably-doing-it-wrong-60373a38d2aa#fb9e

In particular I've linked to the section on [`httptest`](https://pkg.go.dev/net/http/httptest) since it's built-in, but there are other approaches to consider.

Reply:
The overall process to get the device successfully created is when we the device is  provisioned and active which takes more time. For any resource create via terraform provider, which take time to be provisioned, we need to consider the actual `wait` time for that resource to be in desired state and not the API call itself. Most of the time API calls will return in no time. Reference for discussion [here](https://github.com/equinix/terraform-provider-equinix/pull/356#issuecomment-1679316443)

Checked for aws provider [here](https://github.com/hashicorp/terraform-provider-aws/blob/a4393f91f2b25b915a0780ed532b83f2bc68ff2b/internal/service/ec2/ec2_instance.go#L1007) and they have been they consume the timeout while waiting the aws instance to be created. Also checked k8s provider where the api itself is taking and respecting the context with timeout

We can use the terraform helper provided pattern for wait methods as this. As what most of the other providers are also using
https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/retry/state.go#L57

Reply:
The main problem happening with this timeout test is, in the test we provide a config to terraform to create a device with a timeout of very small `5s`. The test return timeout error which is expected and then test tries to destroy the resource. But the issue is equinix apis doesnt allow the device to be destroyed when they are in provisioning state. There's no way to tell terraform to wait and then destroy.

Also, since we are anyway not considering the API call time in our timeout waits, it can be mocked